posts/integrating-civic-into-a-static-serverless-website-part-2-of-2.md

```diff
@@ -0,0 +1,443 @@
+
+We will be using [Civic's server side SDK](https://www.npmjs.com/package/civic-sip-api) which comes as an npm package. First, we need to setup a Civic client for sending tokens. This can be done outside of the function handler itself because multiple calls could use the same client without issue. Importing `jsonwebtoken` is just used for debugging the incoming tokens and is not required for this particular authorizer. Here is the start of `auth.js`:
+
+```javascript
+'use strict';
+
+const civicSip = require('civic-sip-api');
+const jwt = require('jsonwebtoken');
+
+const civicClient = civicSip.newClient({
+    appId: process.env.CIVIC_APP_ID,
+    appSecret: process.env.CIVIC_APP_SECRET,
+    prvKey: process.env.CIVIC_PRIVATE_SIGNING_KEY,
+});
+
+exports.customCivicAuthorizer = async event => {};
+```
+
+We are going to use environment variables for injecting app secrets and the app id. This follows best practices and Serverless can easily pull these values from our AWS Parameter Store. The asynchronous function `customCivicAuthorizer` will be used as the actual Lambda function handler. In `serverless.yml`, we need to link the Parameter Store values to the function variables.
+
+```yaml
+...
+
+functions:
+  ...
+  custom-civic-authorizer:
+    handler: functions/auth.customCivicAuthorizer
+    environment:
+      CIVIC_APP_ID: ${ssm:/tylerjohnhaden/civic-color/APP_ID}
+      CIVIC_APP_SECRET: ${ssm:/tylerjohnhaden/civic-color/APP_SECRET~true}
+      CIVIC_PRIVATE_SIGNING_KEY: ${ssm:/tylerjohnhaden/civic-color/PRIVATE_SIGNING_KEY~true}
+```
+
+Add the above environment variables. Serverless allows us to add them to the [individual functions, or to the whole stage](https://serverless.com/framework/docs/providers/aws/guide/variables/) but we can improve security by limiting the scope of the app secrets to just the authorizer. The `ssm` links to AWS System Manager Parameter Store, and the `~true` tells Serverless to [use your local credentials to decrypt vis KMS](https://hackernoon.com/you-should-use-ssm-parameter-store-over-lambda-env-variables-5197fc6ea45b).
+
+Let us continue with defining the handler. According to [AWS's authorizer documentation](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-lambda-authorizer-input.html), the Lambda authorizer input `event` will have the following pattern:
+
+```json
+{
+    "type": "TOKEN",
+    "authorizationToken": "<caller-supplied-token>",
+    "methodArn": "arn:aws:execute-api:<regionId>:<accountId>:<apiId>/<stage>/<method>/<resourcePath>"
+}
+```
+
+Once we get the token from the input, we can pass it directly to Civic. If this fails, we assume Civic rejected the token and return a `401 Unauthorized` response. Without adding CORS headers to this API response, the browser will actually throw a cross-origin error instead of a 401. This is fixed by setting default 4XX response CORS in API Gateway, which we can do using our Serverless config but more on this later.
```

Comment:
```suggestion
Once we get the token from the input, we can pass it directly to Civic. If this fails, we assume Civic rejected the token and return a `401 Unauthorized` response. Without adding CORS headers to this API response, the browser will actually throw a cross-origin error instead of a 401. This is fixed by setting default 4XX response CORS in API Gateway, which we can do using our Serverless config. More on this later.
```